Simplify _Q_opt in parmest with block scenario structure.

sscini opened the pull request:
<!-- ##################################################################### -->
<!-- PLEASE READ BEFORE OPENING THIS PULL REQUEST -->

<!-- All changes must adhere to PEP8 standards as enforced by Black. -->
<!-- If your changes do NOT adhere, the test suite will fail. -->
<!-- Please read our Contributing guide for instructions on how to apply these standards. -->
<!-- Contributing Guide: https://pyomo.readthedocs.io/en/stable/contribution_guide.html -->
<!-- ##################################################################### -->

<!-- DO NOT DELETE OR IGNORE THIS TEMPLATE. Failing to adhere to this template and provide the necessary information may lead to your Pull Request being closed without consideration. -->

## Fixes # .

## Summary/Motivation:
_Q_opt, the main model building and solving function within parmest, is still heavily dependent on MPISPPY, and the code is becoming outdated and difficult to interpret. The goal of this PR is to redesign _Q_opt to work without this dependence, retaining all the current functionality

## Changes proposed in this PR:
-_Q_opt redesign, using a block structure, with minimal changes needed to functions that utilize it.

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

sscini committed:
First pass at code redesign, still need to figure out more

sscini committed:
Merge branch 'Pyomo:main' into Q_opt-redesign

sscini committed:
Added comments where I have question

sscini committed:
Merge branch 'Q_opt-redesign' of https://github.com/sscini/pyomo into Q_opt-redesign

sscini committed:
Got preliminary _Q_opt simple working with example!

sscini committed:
Ran black

adowling2 reviewed (commented):
Nice progress

sscini committed:
Changed name to _Q_opt_blocks

sscini committed:
Update parmest.py

sscini committed:
Ran black

sscini committed:
Added in case for bootlist, works with example

sscini committed:
Ran black

sscini commented:
@adowling2 @djlaky Added case for bootstrap, now works with theta_est, theta_est_bootstrap, and theta_est_leaveNout if I replace _Q_opt with _Q_opt_blocks. Please review. 

I am aware of the duplicated code, but when I attempted to unify them using n_scenarios = len(self.exp_list) or len(bootlist), I am getting an error for bootstrap I am still working out.

Would we want to fully transition to using theta_est for obj and theta, and cov_est for covariance and add an error/warning? Or should I make it work to calculate cov if calc_cov=True?

Thanks!

sscini committed:
Simplified structure, ran black

sscini commented:
@adowling2 @djlaky Should I tag larger team on this? Please review when available. Thanks!

sscini committed:
Removed _Q_opt, and replicate functions, only using _Q_opt_blocks

sscini committed:
Ran black on mac

sscini committed:
Revert "Removed _Q_opt, and replicate functions, only using _Q_opt_blocks"

This reverts commit 1aea99f4f2ea82a46d2b62459a67cd30693e78a0.

sscini committed:
Added testing statement

sscini committed:
Ran black

sscini committed:
Merge branch 'Pyomo:main' into Q_opt-redesign

sscini committed:
Made small design changes,  in progress, ran black.

sscini committed:
Merge branch 'Q_opt-redesign' of https://github.com/sscini/pyomo into Q_opt-redesign

sscini committed:
Progress made on objective_at_theta_blocks, unfinished.

sscini committed:
Merge branch 'Pyomo:main' into Q_opt-redesign

sscini committed:
Merge branch 'Q_opt-redesign' of https://github.com/sscini/pyomo into Q_opt-redesign

sscini committed:
Added notes for design meeting 01/12/26

sscini committed:
Merge branch 'Pyomo:main' into Q_opt-redesign

sscini committed:
Removed answered reviewer question, attempted adding covariance

sscini committed:
Added assertions for cov_n

sscini committed:
Finished implementing covariance

sscini committed:
Added functional return values argument

sscini committed:
Ran black

sscini committed:
Corrected extraction for unknown parameters

sscini committed:
Initial attempt at objective_at_theta_blocks

sscini committed:
Working out bugs in _Q_at_theta implement. In progress.

sscini committed:
Corrected obj_at_theta_blocks

sscini committed:
Removed _Q_opt, commented out _Q_at_theta, ran black

sscini committed:
Fixed for loop issue

sscini committed:
Removed fix_theta from create_parm_model

sscini committed:
Renamed objective_at_theta_blocks, ran black.

sscini committed:
Removed mentions of _Q_opt_blocks

sscini committed:
Changed back to get_labeled_model in _cov_at_theta()

sscini commented:
@adowling2 @slilonfe5  Ready for review. All functionalities should be working. Two small things to discuss, but relating to covariance calculation in _cov_at_theta. Thank you!

sscini committed:
Added notes in unused files.

sscini committed:
Removed _Q_at_theta and objective_at_theta

sscini committed:
Added comments for reviewers, ran black.

sscini committed:
Corrected count_total_experiments to divide by # outputs.

sscini committed:
Ran black.

sscini committed:
Undo change to count_total_experiments.

sscini committed:
Update mpi_utils.py

sscini committed:
Switched unknown_params to Vars

sscini committed:
Fixed number for cov_n, still need to adjust counting function

sscini committed:
Added question for reviewers

sscini committed:
Ran black

sscini committed:
Changed retrieval of variables for ind_red_hes.

sscini committed:
Added note related to count_total_experiments, commented out assertion.

sscini committed:
Ran black

sscini committed:
Removed dependence on cov_est for theta_est(), added bool for len(exp_list)

cov_est() needs the experiment class to have variables for params, which makes a few tests fail.
Was failing tests with one experiment.

sscini committed:
Merge branch 'main' into Q_opt-redesign

sscini committed:
Ran black

sscini committed:
Attempted adding support for indexed vars

adowling2 commented on pyomo/contrib/parmest/examples/reaction_kinetics/simple_reaction_parmest_example.py:
```diff
@@ -44,6 +44,8 @@ def simple_reaction_model(data):
     model.x2 = Param(initialize=float(data['x2']))
 
     # Rate constants
+    # @Reviewers: Can we switch this to explicitly defining which parameters are to be
```
Yes, I think this example could/should be reworked to fully embrace the Experiment class.

adowling2 commented on pyomo/contrib/parmest/examples/reactor_design/parameter_estimation_example.py:
```diff
@@ -36,9 +36,10 @@ def main():
 
     # Parameter estimation with covariance
-    obj, theta, cov = pest.theta_est(calc_cov=True, cov_n=17)
+    obj, theta, cov = pest.theta_est(calc_cov=True, cov_n=19)
```
Just to satisfy my curiosity, is 17 incorrect and 19 correct? Is this a mistake in the example that is being fixed?

sscini replied:
From the data in the csv file, it should be 19. There are 19 experiments and 4 experiment outputs. The experiment counter (_count_total_experiments) is currently returning 76, which is why I have mentioned making it more robust and commented out the related assertion at this time.

adowling2 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -967,201 +973,286 @@ def TotalCost_rule(model):
         return parmest_model
 
+    # @Reviewers: Is this needed? Calls create_parmest_model above.
     def _instance_creation_callback(self, experiment_number=None, cb_data=None):
```
Is this only for the mpi-sppy deprecated interface?

adowling2 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -967,201 +973,286 @@ def TotalCost_rule(model):
-        outer_cb_data["callback"] = self._instance_creation_callback
-        if ThetaVals is not None:
-            outer_cb_data["ThetaVals"] = ThetaVals
         if bootlist is not None:
```
Is `bootlist` for bootstrapping? Suggestion: add more comments to explain the logic here.

adowling2 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -967,201 +973,286 @@ def TotalCost_rule(model):
+                if ThetaVals:
+                    # Set theta values in the block model
+                    for name in self.estimator_theta_names:
+                        if name in ThetaVals:
```
For example, we could require `ThetaVals` be a list with the same length as the `unknown_parameters` suffix in the annotated model (Experiment class). The user would be responsible for ensuring the list is in the correct order.

slilonfe5 replied:
@sscini I can't see where you defined `self.estimator_theta_names`. Maybe you did, maybe it's just me who can't find it.

adowling2 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -967,201 +973,286 @@ def TotalCost_rule(model):
+        theta_estimates = {}
+        # Extract theta estimates from first block
+        for name in self.estimator_theta_names:
+            theta_estimates[name] = pyo.value(getattr(model.exp_scenarios[0], name))
```
Is there a more robust way to do this besides `getattr`? What do we do in Pyomo.DoE to access variables associated with specific suffixes?

adowling2 reviewed (changes requested):
@sscini Here is my first pass at the review. A handful of places would benefit from more verbose comments to explain the logic or assumptions.

I wonder if there is a better way besides `getattr` to access variable values. I am thinking how we access values in Pyomo.DoE for variables that are in a suffix.

@slilonfe5, please take a look. I think this would benefit from a conversation with you to integrate the different methods to estimate the parameter covariance.

sscini commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -967,201 +973,286 @@ def TotalCost_rule(model):
+            var_values = pd.DataFrame(var_values)
+
+        # Calculate covariance if requested using cov_est()
+        if calc_cov is not NOTSET and calc_cov:
```
[Addresses next few comments]. Need to confirm with @blnicho and team which tests need to be kept and enforced with these changes. If we call cov_est or _cov_at_theta here, then we are restraining the user to only using variables for unknown parameters, which makes multiple of the existing tests and examples fail. The inverse reduced hessian method was what @slilonfe5 implemented to calculate the covariance if a user still wishes to use theta_est for covariance. Previous conversations indicated these features should be preserved from my understanding.
Will also add more comments

sscini commented:
Notes to discuss in 01/20/26 meeting:

Does the use of parameters and indexed parameters still need to be preserved? 
The _count_total_experiments function: Suggestions for path forward.

sscini committed:
Ran black

sscini committed:
Addressed some review comments.

slilonfe5 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -344,6 +347,9 @@ def _get_labeled_model(experiment):
 
 
+# Need to make this more robust. Used in Estimator class
+# Has issue where it counts duplicate data if multiple non-unique outputs
```
> 

@sscini suggested code to fix the issue. I checked, and this works better.

slilonfe5 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -967,201 +973,379 @@ def TotalCost_rule(model):
-        calc_cov=NOTSET,
-        cov_n=NOTSET,
-    ):
+    def _create_scenario_blocks(self, bootlist=None, ThetaVals=None, fix_theta=False):
```
@sscini I think you need the `calc_cov` and `cov_n` arguments. It allows users who use `theta_est` for covariance calculation to continue doing so. If the user doesn't pass the `calc_cov` and `cov_n` arguments in `theta_est`, then `calc_cov=NOTSET` and `cov_n=NOTSET`.

slilonfe5 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -967,201 +973,379 @@ def TotalCost_rule(model):
+                    # Set theta values in the block model
+                    for name in self.estimator_theta_names:
+                        if name in ThetaVals:
+                            var = getattr(parmest_model, name)
```
Can you rename this to something else other than `var`, e.g., `theta_var`

slilonfe5 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -967,201 +973,379 @@ def TotalCost_rule(model):
+                                for index in var:
+                                    val = ThetaVals[name][index]
+                                    var[index].set_value(val)
+                                    if fix_theta:
```
I don't understand why we are fixing theta? Isn't theta what we are solving for in the minimization problem?

sscini commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -967,201 +973,286 @@ def TotalCost_rule(model):
-                    return obj_val, theta_vals, var_values, cov
-                elif calc_cov is NOTSET or not calc_cov:
-                    return obj_val, theta_vals, var_values
+            solve_result, inv_red_hes = (
```
@slilonfe5 I had this implemented. Main concern here was this caused model variants tests to fail (e.g. Params, indexedParams) as we discussed. I will look at extracting the values from the parmest model and review the change with you to fix this.

slilonfe5 replied:
@sscini I see. You are right. This should be a very easy update. We will discuss tomorrow

sscini committed:
Merge branch 'Pyomo:main' into Q_opt-redesign

sscini committed:
Added Shammah fix for exp count

sscini committed:
Updates to address comments.

sscini committed:
Addressed some comments, simplified scenarios

sscini committed:
Replaced getattr with suffix calls.

sscini committed:
Updated, ran black.

sscini committed:
Noted failing tests, currently 15

sscini committed:
Removed old comment during dev

sscini committed:
Fixed scenario count issue, ran black.

sscini committed:
Added else statement in cov calc

sscini committed:
Update test_parmest.py

sscini committed:
Update parmest.py

sscini committed:
Update simple_reaction_parmest_example.py

sscini committed:
Added measurement error to reactor_design

sscini committed:
Changed to built-in SSE

sscini committed:
Commented out model_initialized

sscini committed:
Merge branch 'Pyomo:main' into Q_opt-redesign

sscini committed:
Merge remote-tracking branch 'refs/remotes/origin/Q_opt-redesign' into Q_opt-redesign

sscini committed:
Remove solver import

sscini committed:
Ran black

sscini committed:
Update test_parmest.py

sscini committed:
Added back option, ran black

sscini committed:
Rearranged _Q_opt for fix_theta

sscini committed:
Ran black

sscini committed:
Adjusting parmest models, in progress

sscini committed:
Added more description, simplified comparison

sscini committed:
Update parameter_estimation_example.py

sscini committed:
Ran black

sscini committed:
Adjusted if statement

sscini committed:
Removed answered questions

sscini committed:
Merge branch 'Pyomo:main' into Q_opt-redesign

sscini committed:
Update parmest.py

sscini committed:
Fixed models in variants test

sscini committed:
Merge branch 'Pyomo:main' into Q_opt-redesign

sscini committed:
Merge remote-tracking branch 'refs/remotes/origin/Q_opt-redesign' into Q_opt-redesign

sscini committed:
Update test_parmest.py

sscini committed:
Update parmest.py

sscini committed:
Update parmest.py

sscini committed:
Temporary remove failing test.

sscini committed:
Fixed experiment counter

sscini committed:
Modified testing

sscini committed:
Removed extra print statements

sscini committed:
Updated error message.

sscini committed:
Adjusted experimental counter

Did not work for multi-index like in PDEs. This addresses that.

sscini committed:
Update test_examples.py

adowling2 commented on pyomo/contrib/parmest/examples/reaction_kinetics/simple_reaction_parmest_example.py:
```diff
@@ -55,21 +55,21 @@ def simple_reaction_model(data):
     model.k.fix()
 
     # ===================================================================
```
Can all of this commented-out code just get deleted?

adowling2 commented on pyomo/contrib/parmest/tests/test_parmest.py:
```diff
@@ -901,16 +898,16 @@ def check_rooney_biegler_results(self, objval, cov):
         self.assertAlmostEqual(objval, 4.3317112, places=2)
         self.assertAlmostEqual(
-            cov.iloc[asymptote_index, asymptote_index], 6.30579403, places=2
+            cov.iloc[asymptote_index, asymptote_index], 6.155892, places=2
```
Why are these numbers changing? I want to at least document the reason for the change in this PR.

adowling2 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -971,197 +988,308 @@ def _instance_creation_callback(self, experiment_number=None, cb_data=None):
+        theta_vals : dict, optional
+            Dictionary of theta values to set in the model. If None, use default values from experiment class.
+            Default is None.
+        solver : str, optional
```
Why not just let the user pass in a Pyomo solver object? If they pass in a string, we can create an Ipopt solver object.

adowling2 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -971,197 +988,308 @@ def _instance_creation_callback(self, experiment_number=None, cb_data=None):
+            # Check cov_n argument is set correctly
+            # Needs to be provided
+            assert cov_n is not NOTSET, (
+                "The number of data points 'cov_n' must be provided to calculate "
```
Is this something we can infer from the provided data?

sscini replied:
Yes, we do infer it from the provided data (see _count_total_experiments()), but from the previous implementation it was required for the user to provide both calc_cov = True, and a value for cov_n. The value provided for cov_n is not currently being used, other than to check length in comparison to number of parameters. 

Goal was to not modify current user experience. Is the recommendation to remove the requirement?

adowling2 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -1186,11 +1314,18 @@ def _cov_at_theta(self, method, solver, step):
         if method == CovarianceMethod.reduced_hessian.value:
```
What does this method implicitly assume? I think it assumes the user has specified the objective as the log-likelihood function plus a constant.

adowling2 reviewed (changes requested):
Please see the comments below. There are a few places where I think we want to double check the math. @slilonfe5 and @smondal13, can you also please take a look.

sscini commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -1201,49 +1336,49 @@ def _cov_at_theta(self, method, solver, step):
             )
 
             self.inv_red_hes = inv_red_hes
+        else:
```
(@slilonfe5 confirm please) inv_red_hessian uses the ef_nonants, or the unknown parameter variables from the extended form model, and needs to be run after parameter estimation. _k_aug and finite difference use the model directly from the Experiment class.

'automatic_differentiation_k_aug' and 'finite_difference' are after the else.

sscini committed:
Addressed comments

sscini commented on pyomo/contrib/parmest/examples/reactor_design/reactor_design.py:
Example working normally. Intention was to update the examples in a separate PR. Following recommendation, I will complete now in this PR.

sscini committed:
Merge branch 'Pyomo:main' into Q_opt-redesign

sscini commented on pyomo/contrib/parmest/examples/reactor_design/reactor_design.py:
```diff
@@ -24,15 +24,11 @@ def reactor_design_model():
-    )  # min^-1
-    model.k3 = pyo.Param(
-        initialize=1.0 / 6000.0, within=pyo.PositiveReals, mutable=True
+    # Rate constants, make unknown parameters variables
```
The experiment class is defined after the model definition. Would the preference be to move this into create_model()? The example is showing use of a mutable parameter for model simulation, which I am not sure how this would be updated to use the experiment class.

sscini committed:
Update simple_reaction_parmest_example.py

sscini committed:
Merge branch 'Pyomo:main' into Q_opt-redesign

sscini committed:
Merge branch 'Pyomo:main' into Q_opt-redesign

sscini committed:
Merge branch 'Pyomo:main' into Q_opt-redesign

sscini committed:
Updated interface, added block scenario tests.

sscini committed:
Merge branch 'Pyomo:main' into Q_opt-redesign

adowling2 reviewed (approved):
I think this is ready to merge after review from @blnicho or @mrmundt

sscini committed:
Merge branch 'main' into Q_opt-redesign

sscini committed:
Merge branch 'main' into Q_opt-redesign

sscini committed:
Moved new tests into the main parmest testing file.

sscini committed:
Ran black

blnicho reviewed (commented):
@sscini I'm still going through `parmest.py` but here are my review comments so far on the rest of the files. 

In response to your specific review questions

> 1. How to handle the import structure that only applies to the deprecated interface
    
Could you leave a comment pointing to the "import structure" you're referring to. I'm not following. By `deprecated interface` do you mean the ParmEst interface before the Experiment class or the old interface for calculating the covariance?
> 2. Are there any changes needed in the new implementation I did not consider?
    
Not sure yet. 
> 3. Now that this will be Pyomo 6.10.1, should I remove deprecation warnings for covariance structural change? (separate theta_est and cov_est)
    
Yes, I think you can remove the deprecated covariance functionality.
> 4. Can elements from the new interface be updated into the deprecated and the old structural utility files be removed?
    
Are `create_ef.py` and `mpi_utils.py` the structural utility files you're referring to? I think we should leave these files for now and let's address the rest of the mpisppy infrastructure in a separate PR. Could you be more specific with which "elements" and which deprecated interface you're referring to?

sscini commented:
@blnicho Thank you very much for your comments so far!

I will add a specific comment in the code for the import clarification.

Can elements from the new interface be updated into the deprecated and the old structural utility files be removed? Are create_ef.py and mpi_utils.py the structural utility files you're referring to? I think we should leave these files for now and let's address the rest of the mpisppy infrastructure in a separate PR. Could you be more specific with which "elements" and which deprecated interface you're referring to?

Leaving create_ef.py and mpi_utils makes sense. 

To clarify this: I was referring to the _DeprecatedEstimator class as the deprecated interface. I updated _Q_opt in the Estimator class to use blocks and combined it with _Q_at_theta. The _DeprecatedEstimator still uses both _Q_opt and _Q_at_theta, which rely on the _experiment_instance_creation_callback function (line 95), and have separate booleans to connect to either mpisppy, or create_ef files. 

If I updated the _DeprecatedEstimator to use my _Q_opt implementation, since we previously discussed that the main point of the deprecated estimator is to preserve old user facing functions, then _experiment_instance_creation_callback, and the mentioned files could be removed. The main deprecated function to preserve now would be the use of theta_est for covariance calculations. Otherwise, I tried my best to avoid the user interface in this PR.

Hopefully this clarifies, but please let me know if more information is still needed. As you mentioned, this will most likely be done in its own PR.

sscini committed:
Update test_parmest.py

sscini committed:
Merge branch 'main' into Q_opt-redesign

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -354,14 +357,24 @@ def _count_total_experiments(experiment_list):
+        first_parent_indices = [
+            v for v in output_vars.keys() if v.parent_component() is first_parent
+        ]
+        total_data_points += len(first_parent_indices)
```
This implementation seems fragile to me. What if the experiment outputs are in a different order? What if the first experiment output component is indexed by data point (i.e. time) and another index? What if you're trying to combine data from experiments with different outputs?

Is this calculating the `n` value needed for doing the covariance calculation or what is this value being used for?

sscini replied:
Added assumptions in doc string for function including assuming same time lengths for data sets, and assuming only indexing by time or time is first index. Please let me know if additional changes are needed here.

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -969,197 +982,297 @@ def _instance_creation_callback(self, experiment_number=None, cb_data=None):
-        cov_n=NOTSET,
-    ):
+    def _create_scenario_blocks(self, bootlist=None, theta_vals=None, fix_theta=False):
+        # Create scenario block structure
```
```suggestion
```
This comment can be removed now that there is a docstring

sscini committed:
Merge branch 'Pyomo:main' into Q_opt-redesign

sscini committed:
Removed covariance functionality from theta_est, in progress

sscini committed:
Update simple_reaction_parmest_example.py

sscini committed:
Update simple_reaction_parmest_example.py

sscini committed:
Update parameter_estimation_example.py

sscini committed:
Adjusted tests, ran black

sscini committed:
Update test_parmest.py

sscini committed:
Delete scenarios.csv

blnicho committed:
Merge branch 'main' into Q_opt-redesign

sscini committed:
Merge branch 'Pyomo:main' into Q_opt-redesign

sscini committed:
Merge branch 'Pyomo:main' into Q_opt-redesign

sscini commented on pyomo/contrib/parmest/parmest.py:
Leaving this as is for this PR, unless further recommendations on revising deprecatedEstimator. Currently all user functions have option to access deprecated interface, so I think this would be removed in a future PR unless it is preferred to alter deprecated tests to not use these functions.  Fully removing now would cause new test failures.

sscini commented:
@blnicho Please share remaining comments mentioned for PR and parmest.py when available. Thank you!

sscini committed:
Merge branch 'main' into Q_opt-redesign

sscini committed:
Merge branch 'main' into Q_opt-redesign

sscini committed:
Merge branch 'main' into Q_opt-redesign

sscini committed:
Merge branch 'main' into Q_opt-redesign

sscini committed:
Merge branch 'main' into Q_opt-redesign

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -354,14 +357,24 @@ def _count_total_experiments(experiment_list):
+        # 1. Identify the first parent component of the experiment outputs
+        output_vars = experiment.get_labeled_model().experiment_outputs
+
+        # 1. Identify the first parent component
```
You have two "# 1" comments here

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -969,197 +982,260 @@ def _instance_creation_callback(self, experiment_number=None, cb_data=None):
+
+        for name in expanded_theta_names:
+            template_theta_var = template_model.find_component(name)
+            parent_theta_var = model.parmest_theta[name]
```
Using names to find equivalent theta components across the different experiment models kind of defeats the purpose of the `unknown_parameters` Suffix. I think a better solution might be to enforce that only unindexed components, or "ComponentData",  are allowed in the `unknown_parameters` Suffix. Then the Suffix keys can be used directly and all these extra steps expanding potentially indexed things goes away. If this is too restrictive then in `_expand_indexed_unknowns` we could return a ComponentList or list of cuids instead of names.

sscini replied:
@blnicho @adowling2 Made attempts, but more difficult than initially anticipated to swap for component uids and enforce on user to not use suffixes. Checking what is already enforced in checks on suffix objects to avoid adding unneeded tests.  In progress, but would the current solution work for this PR and a deprecation warning be needed to tell user not to put indexed components in the unknown_parameters? Would appreciate thoughts.

sscini replied:
@blnicho After discussion with @slilonfe5 and @adowling2, we agreed it will be valuable to enforce non-indexed suffix values. For this release, I implemented a change to use both componentUIDs and names, preventing restriction on the user. Storing names is valuable for objective_at_theta or future multistart implementations where a dataframe is provided with the names for each column, and are checked in that function. This is in line with the current tests from test_parmest.py for different supported model variants.

In future we will enforce the expanded suffix values, similar to time indexed components in ParmEst or DoE,  as best practice, and deprecate the use of indexed suffix values for unknown_parameters. Thanks!

sscini committed:
Merge branch 'Pyomo:main' into Q_opt-redesign

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -969,197 +982,260 @@ def _instance_creation_callback(self, experiment_number=None, cb_data=None):
+            raise ValueError("At least one scenario is required to build the EF model.")
+
+        # Create indexed block for holding scenario models
+        model.exp_scenarios = pyo.Block(range(self.obj_probability_constant))
```
Why not index this block by the `scenario_numbers` list you created above? Maybe add `scenario_numbers` to the model as a Pyomo `Set` so that you have easy access to it later?

sscini replied:
Switched to index by scenario numbers as suggested. Stored in an indexed parameter instead of set because bootlist (theta_est_bootstrap, leaveNout, etc.) can have duplicate entries.

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -969,197 +982,260 @@ def _instance_creation_callback(self, experiment_number=None, cb_data=None):
-                        tee=self.tee,
+        # Set the number of experiments to use, either from bootlist or all experiments
+        scenario_numbers = (
+            list(bootlist) if bootlist is not None else list(range(len(self.exp_list)))
```
Why do you cast bootlist to a list here? Shouldn't it already be a list based on the type mentioned in the docstring?

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -969,197 +982,260 @@ def _instance_creation_callback(self, experiment_number=None, cb_data=None):
+
+        # Create indexed block for holding scenario models
+        model.exp_scenarios = pyo.Block(range(self.obj_probability_constant))
+        for i, experiment_number in enumerate(scenario_numbers):
```
If you index the block by `scenario_numbers` directly then you don't have to worry about this mapping between the list index and the experiment number.

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -969,197 +982,260 @@ def _instance_creation_callback(self, experiment_number=None, cb_data=None):
+                    child_theta_var = model.exp_scenarios[i].find_component(name)
+                    model.theta_link_constraints.add(
+                        child_theta_var == parent_theta_var
                     )
```
You spend a lot of time here finding the same component multiple times. You could make this more efficient by waiting to set the value of the `child_theta_var` until the second for-loop. Use the first for-loop only for building the scenario blocks.

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -969,197 +982,260 @@ def _instance_creation_callback(self, experiment_number=None, cb_data=None):
+        theta_vals=None,
+        fix_theta=False,
+    ):
+        '''
```
Please use double quotes for docstrings
```suggestion
        """
```

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -969,197 +982,260 @@ def _instance_creation_callback(self, experiment_number=None, cb_data=None):
+        fix_theta=False,
+    ):
+        '''
+        Making new version of _Q_opt that uses scenario blocks, similar to DoE.
```
Could you make this method summary a bit more verbose. It currently reads like a commit message rather than a description of what the method is doing. You don't need to refer to it as a "new version".

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -969,197 +982,260 @@ def _instance_creation_callback(self, experiment_number=None, cb_data=None):
-                    cov, index=theta_vals.keys(), columns=theta_vals.keys()
-                )
+    # Redesigned _Q_opt method using scenario blocks, and combined with
+    # _Q_at_theta structure.
```
These comments can likely be removed.
```suggestion
```

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -969,197 +982,260 @@ def _instance_creation_callback(self, experiment_number=None, cb_data=None):
-            theta_vals = pd.Series(theta_vals)
+        Steps:
+        1. Load model - parmest model should be labeled
+        2. Create scenario blocks (biggest redesign) - clone model to have one per experiment
```
```suggestion
        2. Create scenario blocks - clone model to have one per experiment
```

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -969,197 +982,260 @@ def _instance_creation_callback(self, experiment_number=None, cb_data=None):
+        calc_cov : bool, optional
+            If True, calculate covariance matrix of estimated parameters. Default is NOTSET.
+        cov_n : int, optional
+            Number of data points to use for covariance calculation. Required if calc_cov is True. Default is NOTSET.
```
I think we should mention in the docstring that these arguments are deprecated similar to the description in `theta_est`

sscini replied:
No longer arguments for _Q_opt. I removed from docstring instead. Please let me know if I should put back.

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -969,197 +982,260 @@ def _instance_creation_callback(self, experiment_number=None, cb_data=None):
+            obj_value : float
+                Objective value at fixed parameter values.
+            theta_estimates : dict
+                Dictionary of fixed parameter values.
```
It seems kind of weird that the return type of `theta_estimates` changes depending on the value of `fix_theta`. Why is this necessary?

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -969,197 +982,260 @@ def _instance_creation_callback(self, experiment_number=None, cb_data=None):
+        # Print model if in diagnostic mode
+        if self.diagnostic_mode:
+            print("Parmest _Q_opt model with scenario blocks:")
+            model.pprint()
```
You should use a logger here instead of printing directly to stdout

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -969,197 +982,260 @@ def _instance_creation_callback(self, experiment_number=None, cb_data=None):
+                # )
+                # Unless infeasible, update worst_status
+                if worst_status != pyo.TerminationCondition.infeasible:
+                    worst_status = status
```
Why do you obscure an "infeasible" termination condition here? Why not just return the termination condition directly?

sscini replied:
Addressed with load_solutions as discussed. It seems the reasoning for neglecting infeasible is because the solve result is empty? Still working now and can make more robust either based on feedback or future PRs where deprecated interface fully removed.

sscini committed:
Merge branch 'main' into Q_opt-redesign

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -969,197 +982,260 @@ def _instance_creation_callback(self, experiment_number=None, cb_data=None):
+        if return_values is not None and len(return_values) > 0:
+            var_values = []
+            # In the scenario blocks structure, exp_scenarios is an IndexedBlock
+            exp_blocks = self.ef_instance.exp_scenarios.values()
```
This is the only place you use `self.ef_instance` in this method instead of `model` so I suggest making this change for consistency
```suggestion
            exp_blocks = model.exp_scenarios.values()
```

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -1181,14 +1257,20 @@ def _cov_at_theta(self, method, solver, step):
-                ind_vars.append(Var)
-            # calculate the reduced hessian
+            for name in self.ef_instance._parmest_theta_names:
+                var = self.ef_instance.parmest_theta[name]
+                ind_vars.append(var)
+
```
I suspect you can simplify this to:
```suggestion
            # retrieve the independent variables (i.e., estimated parameters)
            ind_vars = list(self.ef_instance.parmest_theta.values())
```

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -1973,65 +1867,73 @@ def objective_at_theta(self, theta_values=None, initialize_parmest_model=False):
+                "The `initialize_parmest_model` option in `objective_at_theta()` is "
+                "deprecated and will be removed in future releases. Please ensure the"
+                "model is initialized within the Experiment class definition.",
+                version="6.9.5",
```
If this is a new deprecation warning introduced in this PR then our standard is to set the version to the current dev version of Pyomo. We'll update it to the next official Pyomo version as part of our release process.
```suggestion
            deprecation_warning(
                "The `initialize_parmest_model` option in `objective_at_theta()` is "
                "deprecated and will be removed in future releases. Please ensure the"
                "model is initialized within the Experiment class definition.",
                version="6.10.1.dev0",
```

blnicho committed:
Merge branch 'main' into Q_opt-redesign

sscini committed:
Merge branch 'Pyomo:main' into Q_opt-redesign

sscini committed:
Update pyomo/contrib/parmest/parmest.py

Co-authored-by: Bethany Nicholson <blnicho@users.noreply.github.com>

sscini committed:
Addressing PR comments, in progress

sscini committed:
Ran black, tests passing, in progress

sscini committed:
Addressed more comments, ran black, testing in progress

sscini commented:
@blnicho @mrmundt Good morning! I am actively addressing previous comments, but it looks like there may be an issue with documentation tests? I am unsure if the problems are from my changes or something else. If you could please take a look it would be appreciated

blnicho commented:
> @blnicho @mrmundt Good morning! I am actively addressing previous comments, but it looks like there may be an issue with documentation tests? I am unsure if the problems are from my changes or something else. If you could please take a look it would be appreciated

@sscini to me it looks like you have a rogue debugging `pprint` somewhere in your code.

sscini commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -969,197 +1002,261 @@ def _instance_creation_callback(self, experiment_number=None, cb_data=None):
-            elif calc_cov is NOTSET or not calc_cov:
-                return obj_val, theta_vals
+        logger.info("Parmest _Q_opt model with scenario blocks:")
+        logger.info(model.pprint())
```
@blnicho This should probably be changed to something higher like DEBUG? I think this is causing the issue,

slilonfe5 replied:
@sscini The issue is having model.pprint() inside the logger. You only need one logger call. In that logger call, pass the message you want to show to the user (usually supplied as a string).

sscini committed:
Update parmest.py

sscini committed:
Merge branch 'main' into Q_opt-redesign

sscini committed:
Updated logger issue, ran black

sscini committed:
Merge branch 'Q_opt-redesign' of https://github.com/sscini/pyomo into Q_opt-redesign

sscini committed:
Addressed list comment, refactor for indexing by scenarios in progress

sscini committed:
Attempting scenario storage and indexing.

sscini committed:
Update test_parmest.py

sscini committed:
Small comment change

sscini committed:
Attempt at making exp count more robust

sscini committed:
Revert "Attempt at making exp count more robust"

This reverts commit dd5cbb4696b0778d4798417a2c50e29ed753e2b6.

sscini committed:
Attempt at adding termination condition support.

sscini committed:
Test failures address, ran black.

sscini committed:
Resolved issues with experiment count, tests passing

sscini committed:
Merge branch 'main' into Q_opt-redesign

sscini commented:
@blnicho Addressed comments other than enforcing no indexed unknown parameters. Getting more guidance on this from @adowling2 and @slilonfe5, and if you feel is necessary for this PR will include proposed solution by the end of the week. Please review again when available. I believe failing check will pass if rerun. Thank you!

slilonfe5 committed:
Updated the datapoint counting function in parmest.py

slilonfe5 committed:
Fixed string formatting issues in parmest.py file

slilonfe5 committed:
Added assertions to check the experiment outputs

slilonfe5 committed:
Ran black

slilonfe5 committed:
Final polishing for data point counting

slilonfe5 committed:
Merge branch 'main' into Q_opt-redesign

slilonfe5 committed:
Minor string formatting

slilonfe5 committed:
Removed parentheses in assertion strings

slilonfe5 committed:
Merge branch 'main' into Q_opt-redesign

sscini committed:
Addressed inconsistent returns in _Q_opt

sscini committed:
Ran black

sscini committed:
Merge branch 'Pyomo:main' into Q_opt-redesign

sscini committed:
Merge branch 'Q_opt-redesign' of https://github.com/sscini/pyomo into Q_opt-redesign

sscini committed:
Modified to use CUIDs

sscini committed:
Ran black

slilonfe5 committed:
Removed print statements

sscini committed:
Remove redundant tests

sscini committed:
Ran black

mrmundt commented on pyomo/contrib/parmest/tests/test_parmest.py:
```diff
@@ -1416,6 +1418,362 @@ def test_theta_est_with_square_initialization_diagnostic_mode_true(self):
+        solver = mock.Mock()
+        solver.options = {}
+        solver.solve.return_value = solve_result
+        return solver
```
I understand the thought process here from a "pure unit test" perspective of mocking a solver, but I'd argue that it's actually overall stronger for Pyomo to ensure that real solvers work with all tests (unless it's a test against the units of a solver itself). Because solvers are so fundamentally important to the Pyomo ecosystem/infrastructure, mocking one is significantly weaker than just using a real one.

sscini replied:
Thank you for mentioning this. I will remove mock tests and replace with real checks here and in comments below. Will avoid use moving forward. To be candid, was recommended to save compute time by AI, and wanted to ensure future tests did not add a large additional cost to testing as discussed previously for ParmEst and DoE.

mrmundt commented on pyomo/contrib/parmest/tests/test_parmest.py:
```diff
@@ -1416,6 +1418,362 @@ def test_theta_est_with_square_initialization_diagnostic_mode_true(self):
+                        obj, theta = pest._Q_opt()
+                        obj_with_vars, theta_with_vars, var_values = pest._Q_opt(
+                            return_values=["y"]
+                        )
```
Oh this nested mock is really confusing to look at. Why mock _so many_ different objects? See previous comment for more thought-process on mocking in general, but in particular here, this makes me confused as to what this test is even doing.

sscini committed:
Moved deprecated functions down

mrmundt commented on pyomo/contrib/parmest/tests/test_parmest.py:
```diff
@@ -1416,6 +1418,362 @@ def test_theta_est_with_square_initialization_diagnostic_mode_true(self):
+                    ):
+                        obj, theta, status = pest._Q_opt(
+                            theta_vals={"theta": 9.0}, fix_theta=True
+                        )
```
Why make a real scenario block and then mock it later?

sscini committed:
Ran black

mrmundt reviewed (changes requested):
I am confused about the usage of `mock` in the tests.

sscini committed:
Removed print statement

sscini committed:
Addressed mock tests and new failure

sscini committed:
ran black

sscini committed:
Remove mock import

mrmundt commented on pyomo/contrib/parmest/tests/test_parmest.py:
```diff
@@ -1315,10 +1314,11 @@ def test_covariance(self):
         # In this example, this is the number of data points in data_df, but that's
         # only because the data is indexed by time and contains no additional information.
-        n = 60
+        n = 20
```
This was changed to 20; does the comment need to be updated?

sscini replied:
Adjusted, @slilonfe5 please confirm or adjust the language if needed.

slilonfe5 replied:
That is correct Stephen, 20 is the right number

slilonfe5 replied:
This is because all the CA, CB, and CC measurements are collected at the 20 time points. The time points is the number of experiments

sscini committed:
Update test_parmest.py

sscini commented:
@mrmundt Please rerun the one failing check. It failed really early.

slilonfe5 commented:
@blnicho, @mrmundt, I think all of the review comments have been addressed. We would appreciate final comments if there are any, otherwise, I think this looks pretty good.

mrmundt commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -2198,6 +2129,158 @@ def confidence_region_test(
+
+    if "BootList" in outer_cb_data:
+        bootlist = outer_cb_data["BootList"]
+        # print("debug in callback: using bootlist=",str(bootlist))
```
Probably don't need this anymore but not critical to remove

slilonfe5 replied:
@mrmundt This appears to be in the deprecated interface. Should this be done in a future PR?

mrmundt commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -2198,6 +2129,158 @@ def confidence_region_test(
+            "Only one callback signature is supported: "
+            "callback(experiment_number, cb_data) "
+        )
+        """
+        try:
+            instance = callback(scenario_tree_model, scen_name, node_names)
+        except TypeError:  # deprecated signature?
+            try:
+                instance = callback(scen_name, node_names)
+            except:
+                print("Failed to create instance using callback; TypeError+")
+                raise
+        except:
+            print("Failed to create instance using callback.")
+            raise
+        """
```
Also probably not needed?

slilonfe5 replied:
@mrmundt Similar comment here. As this is in the deprecated interface, should this be done in a future PR?

sscini replied:
Thank you @mrmundt and @slilonfe5. In previous discussions, @blnicho and I agreed to use a separate PR to remove unused code and potentially update the deprecated interface. I will try removing this, but it might lead to new test failures in deprecated. Will happily discuss in the dev meeting today.

mrmundt reviewed (approved):
I have two comments about what look like residual comments / blocks, but they aren't critical to remove.

slilonfe5 committed:
Added a few tests to improve coverage and removed duplicated exceptions in parmest.py

slilonfe5 committed:
Merge branch 'main' into Q_opt-redesign

slilonfe5 committed:
Ran black

slilonfe5 committed:
Final polishing of comments in data counting

blnicho committed:
Merge branch 'main' into Q_opt-redesign

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -345,10 +194,113 @@ def _get_labeled_model(experiment):
+    Checks if experiment outputs are scalars (i.e., not indexed) or their indices
+    are local (i.e., strings), e.g., `m.y1`, `m.y2`, `m.C["A"]`, `m.C["B"]`
+    """
+    return index is None or isinstance(index, str)
```
Why is this checking for string indices? Is this assuming that if the experiment outputs are indexed, the index must be a string?

slilonfe5 replied:
@blnicho No. It is just checking for a special case of indexed outputs (called locals) whose index is a string (which is not a data index but an identity index). I will remove these cases from the code

slilonfe5 replied:
Resolved

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -345,10 +194,113 @@ def _get_labeled_model(experiment):
+def _check_index_is_none_or_local(index):
+    """
+    Checks if experiment outputs are scalars (i.e., not indexed) or their indices
+    are local (i.e., strings), e.g., `m.y1`, `m.y2`, `m.C["A"]`, `m.C["B"]`
```
The terminology here is confusing. What do you mean by a "local" index and why does that imply it must be a string? In the past we have referred to indexing sets as explicit (`m.C[t]`) or implicit (`m.b[t].C`) but I'm not sure if that's what you're trying to check for here.

slilonfe5 replied:
Resolved

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -345,10 +194,113 @@ def _get_labeled_model(experiment):
+
+        # check if the output variable is a scalar (e.g., m.y1, m.y2)
+        # or has a local index (e.g., m.C["A"], m.C["B"])
+        if _check_index_is_none_or_local(index):
```
Again, I think using the term "local" here is unclear. Are you just trying to distinguish between scalar and indexed variables?

slilonfe5 replied:
Resolved

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -345,10 +194,113 @@ def _get_labeled_model(experiment):
+
+    Assumptions:
+
+    Experiment outputs can be scaler or local variables
```
Change to "scaler or indexed variables"

sscini replied:
Scaler or Scalar? Double checking before correction

slilonfe5 replied:
Resolved

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -360,19 +312,47 @@ def _count_total_experiments(experiment_list):
+
+        # Case 1 and 2:
+        # scalar outputs such as m.y1, m.y2,...
+        # local outputs such as m.C["A"], m.C["B"],...
```
Change "local" to "indexed". Is the `m.C["A"]` case even valid? Aren't you assuming that if the outputs are indexed, the first indexing set must be the "data" index?

slilonfe5 replied:
@blnicho I will remove the `m.C["A"]` case (i.e., local index case). Yes, we are assuming that if an output is indexed, the first indexing must be the data index

slilonfe5 replied:
Resolved

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -845,6 +825,11 @@ def __init__(
         # boolean to indicate if model is initialized using a square solve
         self.model_initialized = False
 
+        # If self.diagnostic mode is true, then set the logging level to INFO to print
```
Typo: "INFO" should be "DEBUG" based on the implementation

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -983,76 +978,99 @@ def _instance_creation_callback(self, experiment_number=None, cb_data=None):
-        model : ConcreteModel
-            Pyomo model with scenario blocks for parameter estimation. Contains indexed block for
-            each experiment in exp_list or bootlist.
-
```
Why did you delete this more detailed docstring?

sscini replied:
I think I got confused because a lot of the private functions have much shorter docstrings. Added back now

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -1085,12 +1101,20 @@ def _Q_opt(
-        Making new version of _Q_opt that uses scenario blocks, similar to DoE.
+        """
+        _Q_opt method for parameter estimation using an extended form
+        optimization problem with scenario blocks for each experiment.
```
Should "extended form" be "extensive form"?

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -1156,31 +1177,24 @@ def _Q_opt(
+                or len(solve_result.solution) == 0
+            ):
+                theta_payload = dict(theta_vals) if theta_vals is not None else {}
+                return None, theta_payload, termination_condition
```
Why did you use the variable name "theta_payload" here when you use the name "theta_estimates" everywhere else? Also, the third return value here is the termination condition which doesn't match the description of the return values in the docstring.

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -1235,8 +1246,6 @@ def _Q_opt(
         else:
             return obj_value, theta_estimates
```
Every return statement in this function should return the same number of things and the same types matching the description in the docstring.

sscini replied:
I am working to correct the docstring to be more accurate and returns are consistent. 

For the fix_theta cases, I plan to use booleans to show the different instances.

If fix_theta is true: it will always return obj_value, theta_estimates, termination condition

If fix_theta is False and return_values is None: it will return obj_value, theta_estimates

If fix_theta is False and return_values is not None: it will return obj_value, theta_estimates, var_values

In this case the return for var_values is denoted as optional, which is similar to what was done when the options were available for covariance (cov_n and calc_cov). There were previously (current main) either 2, 3, or 4 returns depending on the options selected.

blnicho commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -1192,19 +1206,16 @@ def _Q_opt(
-
-        # Return theta estimates as a pandas Series
-        theta_estimates = pd.Series(theta_estimates)
+            return obj_value, theta_estimates, termination_condition
```
Again you're returning the termination condition as the third return value which doesn't match the docstring.

slilonfe5 committed:
Removed the local case from the datapoint counting

slilonfe5 committed:
Merge branch 'Q_opt-redesign' of https://github.com/sscini/pyomo into Q_opt-redesign

slilonfe5 committed:
Improved code flow in data point counting

sscini committed:
Addressed comments, ran black,

sscini commented:
@blnicho Requested changes are applied, please rerun failing tests as the issues are related to windows installation. Thanks!

sscini committed:
Change to doc string, ran black

sscini commented:
> @blnicho Requested changes are applied, please rerun failing tests as the issues are related to windows installation. Thanks!

This was before I made one more small change. Hopefully new set is all passing.

sscini commented:
@blnicho One bad windows check. Please rerun. Thank you!!

sscini commented:
@blnicho Please let me know if I need to add a few more tests or good as is. Some of the uncovered portion in the patch is due to me moving the deprecated functions down, and will be addressed soon in a PR. Thank you!
